🚀 3단계 - 지하철 구간 관리

domodazzi opened the pull request:
### 작업내용
* 구간 등록 API 작성
* 구간 제거 API 작성
* ControllerAdvice 작성
* 구간 추가,제거 검증을 위한 validator 작성
* 예외 케이스에 대한 인수테스트 작성

### 질문
* 패키지 구조를 정리할 필요를 느끼고 있어요! section과 관련된 패키지를 어떻게 분리해내면 좋을지 고민 중 입니다

오랜만에 작업을 올리는데 얼른 진도 따라가겠습니다!

Commit:
refactor : 리뷰 피드백 반영

- LineTest 클래스 이름 LineAcceptanceTest로 변경
- 테스트 환경 관련 코드 AcceptanceTest 상위 클래스로 이관

Commit:
style : 테스트 코드 클래스 세분화, controller 메소드 오타 수정

Commit:
test : 지하철 구간 관련 인수 테스트 작성

Commit:
feat : 지하철 구간 생성 API 작성

Commit:
fix : section fixture 버그 수정, 오타 수정

Commit:
feat : 지하철 구간 생성 API 수정

- 새로운 노선 생성 시 구간 추가 검증 제외
- 노선 조회 시 상행, 하행을 포함한 모든 역 정보 조회되도록 수정

Commit:
feat : controller advice 추가

Commit:
fix : rest assured test 깨진 것 수정

Commit:
feat : 지하철 구간 제거 API 작성

Commit:
refactor : 지하철 노선의 연관 객체들의 영속성 전이 속성 수정

Commit:
fix : 구간 제거 기능 요구사항 오반영 된 것 수정

- query string 으로 받는 인자가 station ID 인데 section ID로 구현한 부분 수정

junwoochoi commented on src/main/java/subway/line/Line.java:
```diff
@@ -27,6 +32,9 @@ public class Line {
   @Column(nullable = false)
   private int distance;
```
Line의 upStationId, downStationId, distance는 더이상 필요없으니 제거하고 sections를 활용해보시는 방향으로 변경해보시면 어떨까요~?

junwoochoi commented on src/main/java/subway/line/Line.java:
```diff
@@ -27,6 +32,9 @@ public class Line {
   @Column(nullable = false)
   private int distance;
 
+  @OneToMany(mappedBy = "line", fetch = FetchType.LAZY, cascade = CascadeType.PERSIST, orphanRemoval = true)
```
cascade, orphanRemoval 활용해서 line을 통해 section의 라이프사이클에 접근하도록 해주셨네요 👍

junwoochoi commented on src/main/java/subway/line/Line.java:
```diff
@@ -58,9 +66,40 @@ public Long getDownStationId() {
   }
 
+  public void addSection(final Section section) {
+    // 새 라인의 경우 section 검증 제외
+    if (!isNewLine()) {
+      LineValidator.checkSectionForAddition(this, section);
+    }
+
+    this.downStationId = section.getDownStationId();
+    this.distance += section.getDistance();
+    this.sections.add(section);
+    section.registerLine(this);
+  }
+
+  public void removeSection(final Section section) {
+    LineValidator.checkSectionForRemove(this, section);
+
+    this.downStationId = section.getUpStationId();
+    this.distance -= section.getDistance();
+    this.sections.remove(section);
```
section을 추가/제거 할때의 검증을 LineValidator라는 static 클래스에서 하고 있는데요! 아마 최초에 Line도메인이 구간들의 검증에 대한 책임을 갖고 있는 것이 조금 어색하지 않은가 하는 마음에 validator를 추가해주신 것으로 생각됩니다! 
 Sections라는 일급 컬렉션을 만들고 해당 객체가 구간들간의 검증에 대한 책임을 가져보는 것이 응집도면에서는 더 좋을 것 같은데 한번 고민해봐주시면 좋을 것 같습니다~ :)
 
 https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/

domodazzi replied:
좋은 의견 감사합니다! 2주차 미션으로 코드를 옮기면서 적용해보겠습니다 🙇‍♂️

junwoochoi commented on src/main/java/subway/line/LineService.java:
```diff
@@ -51,33 +56,67 @@ public List<LineResponse> findAllLines() {
+
+    line.addSection(section);
+
+    return SectionResponse.from(sectionRepository.save(section));
```
현재 sections의 cascade옵션과 orphanRemoval옵션을 적절히 잘 설정해주셨기 때문에 별도로 sectionRepository를 사용할 필요없이 line을 저장할때 section이 같이 저장될 수 있습니다. 확인 부탁드려요!

junwoochoi commented on src/main/java/subway/line/LineController.java:
```diff
@@ -34,20 +35,42 @@ public ResponseEntity<List<LineResponse>> getLines() {
   }
 
+  @PostMapping("/{id}/sections")
+  public ResponseEntity<SectionResponse> createSection(
```
응답에 sectionResponse가 필요할까요?
구간에 대한 접근은 항상 노선을 통해서만 접근하다보니 굳이 section의 id를 외부에 노출하지 않아도 될 것 같아요 ~! 해당 id가 line의 id라고 오해할 여지가 있어보여서 코멘트 드립니다.

junwoochoi commented on src/main/java/subway/line/LineService.java:
```diff
@@ -51,33 +56,67 @@ public List<LineResponse> findAllLines() {
+    final var line = lineRepository.findById(lineId)
+        .orElseThrow(() -> new BusinessException("노선 정보를 찾을 수 없습니다."));
+
+    final var section = line.getSections().stream()
+        .filter(it -> it.getDownStationId().equals(stationId))
+        .findAny()
+        .orElseThrow(() -> new BusinessException("구간 정보를 찾을 수 없습니다."));
```
비즈니스 로직에 해당하는 부분을 서비스 클래스가 아닌 도메인 클래스에 위치하도록 해볼 수 있을까요?

junwoochoi commented on src/main/java/subway/line/LineService.java:
```diff
@@ -51,33 +56,67 @@ public List<LineResponse> findAllLines() {
+  ) {
+    final var line = lineRepository.findById(lineId)
+        .orElseThrow(() -> new BusinessException("노선 정보를 찾을 수 없습니다."));
+    final var section = new Section(null, upStationId, downStationId, distance, line);
```
section클래스의 생성에 대한 책임도 line내부로 이동해볼 수 있을 것 같아요 :)

junwoochoi commented on src/test/java/subway/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,432 @@
+    @Test
+    void createSectionSuccess() {
+      // given
+      final var upStation = StationFixture.createStation("강남역");
+      final var downStation = StationFixture.createStation("청계산입구역");
+      final var line = LineFixture.createLine("신분당선", "bg-red-600", upStation.getId(), downStation.getId(), 10);
+      final var newStation = StationFixture.createStation("논현역");
```
given 부분이 반복되는데 반복을 줄여보면 좋겠습니다 :)

junwoochoi commented on src/test/java/subway/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,432 @@
+          .map(StationResponse::getId)
+          .collect(Collectors.toList());
+
+      assertThat(stationIds).containsAll(
```
순서까지 검증할 수 있는 containsExactly를 활용하면 더 좋을 것 같습니다! :)

junwoochoi commented on src/test/java/subway/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,432 @@
+      assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+
+      // then
+      // TODO 에러 응답 내용 추가
```
TODO 부분도 추후에는 작성해보시면 좋은 연습이 될 것 같아요~ 
이번 단계에서는 의도적으로 생략한 것으로 이해하고 있겠습니다!

junwoochoi commented on src/test/java/subway/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,432 @@
+          .extract();
+
+      // then
+      assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+
+      // then
+      // TODO 에러 응답 내용 추가
+    }
+
+    /**
+     * When 노선에 구간이 하나 밖에 없다면
+     * Then 에러를 응답한다.
+     */
+    @DisplayName("노선에 구간이 하나 밖에 없다면 삭제할 수 없다.")
+    @Test
+    void deleteSectionFailCanNotDeleteLastSection() {
+      // given
+      final var 청계산입구역 = StationFixture.createStation("청계산입구역");
+      final var 강남역 = StationFixture.createStation("강남역");
+      final var line = LineFixture.createLine("신분당선", "bg-red-600", 청계산입구역.getId(), 강남역.getId(), 10);
+      final var sectionId = 1;
+
+      // when
+      final var response = RestAssured
+          .given().log().all()
+          .queryParam("sectionId", sectionId)
+          .contentType(MediaType.APPLICATION_JSON_VALUE)
+          .when().delete("/lines/{id}/sections", line.getId())
+          .then().log().all()
+          .extract();
+
+      // then
+      assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+
+      // then
+      // TODO 에러 응답 내용 추가
+    }
+
+  }
+
```
예외 케이스에 대한 인수테스트 꼼꼼히 추가 💯

junwoochoi commented on src/main/java/subway/common/ControllerAdvice.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class ControllerAdvice {
+
+  @ExceptionHandler(BusinessException.class)
+  public ResponseEntity<String> handleBusinessException(BusinessException e) {
+    return ResponseEntity
+        .badRequest()
+        .body(e.getMessage());
+  }
+
+  @ExceptionHandler(RuntimeException.class)
+  public ResponseEntity<String> handleRunTimeException(RuntimeException e) {
+    return ResponseEntity
+        .internalServerError()
+        .body("undefined error occurred");
+  }
```
공통 예외처리 👍

junwoochoi reviewed (changes requested):
안녕하세요! 요구사항에 맞추어 미션을 잘 진행해주셨네요 👍 
다음 단계로 넘어가기전에 확인해보면 좋을만한 코멘트 몇가지 남겼으니 확인해 주시고 다시 리뷰 요청 부탁드리겠습니다! 🙇‍♂️

junwoochoi commented:
패키지 구조의 경우, 제가 생각했을 때 `구간`은 `노선`의 하위 도메인에 속한다고 생각해서 현재처럼 `구간`패키지 하위에 속해도 괜찮다고 생각합니다 :)
패키지 하위에 클래스가 지나치게 많게 느껴지신다면, 각 도메인기준으로 패키지를 큰 틀에서 나누고, 해당 패키지 하위에 도메인, 서비스, 컨트롤러 등과 같이 하위 패키지로 한번 더 나눠보셔도 좋을 것 같아요 😄

Commit:
fix : 리뷰 피드백 반영

- 구간 생성 API 필요없는 응답 제거
- 구간 추가 로직에서 save 메소드 호출 부 제거, 영속성 전이를 통해 관리되도록 수정
- line 클래스 distance, upStationId, downStationId 필 제거

domodazzi commented:
꼼꼼하게 봐주셔서 감사합니다 👍  피드백 주신 내용들 반영해봤습니다

Commit:
fix : 리뷰 피드백 반영

- 테스트 가독성 개선

Commit:
chore : 주석 오타 수정

junwoochoi commented on src/main/java/subway/line/LineService.java:
```diff
@@ -119,4 +96,16 @@ public void deleteSection(
+      return Collections.emptyList();
+    }
+
+    List<Long> stationIds = line.getSections().stream()
+        .map(Section::getDownStationId)
+        .collect(Collectors.toList());
+    stationIds.add(0, line.getSections().get(0).getUpStationId());
```
JPA에서 따로 설정을 해주지 않으면 sections가 항상 정렬되서 내려오진 않을 수 있습니다. 현재는 앞에서부터 순서대로만 넣어주고 있지만 추후에 중간에서 넣고 빼고 할때는 유의해보시면 좋을 것 같아요~

junwoochoi commented on src/test/java/subway/LineAcceptanceTest.java:
```diff
@@ -257,12 +255,10 @@ void createSectionSuccess() {
-              downStation.getId(),
-              newStation.getId()
-          )
+      assertThat(stationIds).containsExactly(
```
👍

junwoochoi reviewed (approved):
안녕하세요! 피드백 적용해주시고, 일부 피드백은 2단계 미션 사전에 적용해보시겠다고 하여서 머지하겠습니다. 다음 주차 미션 시작전에 피드백드린 내용들 한번더 확인해봐주시고, 다음 단계도 화이팅 해주세요! 수고많으셨습니다 🙇
